neuralpredictors/layers/readouts/attention.py

```diff
@@ -1,88 +1,96 @@
 
         self.init_noise = init_noise
         if bias:
-            bias = Parameter(torch.Tensor(self.outdims))
-            self.register_parameter("bias", bias)
+            bias_param = Parameter(torch.Tensor(self.outdims))
+            self.register_parameter("bias", bias_param)
         else:
             self.register_parameter("bias", None)
         self.initialize(mean_activity)
 
     @staticmethod
-    def init_conv(m):
-        if isinstance(m, nn.Conv2d):
-            nn.init.xavier_normal_(m.weight.data)
+    def init_conv(m: nn_modules.Module) -> None:
+        if isinstance(m, nn_modules.Conv2d):
+            init.xavier_normal_(m.weight.data)
             if m.bias is not None:
                 m.bias.data.fill_(0)
 
-    def initialize_attention(self):
+    def initialize_attention(self) -> None:
         self.apply(self.init_conv)
 
-    def initialize(self, mean_activity=None):
+    def initialize(self, mean_activity: Optional[Mapping[str, float]] = None) -> None:  # type: ignore[override]
         if mean_activity is None:
             mean_activity = self.mean_activity
         self.features.data.normal_(0, self.init_noise)
         if self.bias is not None:
-            self.initialize_bias(mean_activity=mean_activity)
+            self.initialize_bias(mean_activity=mean_activity)  # type: ignore[no-untyped-call]
         self.initialize_attention()
 
-    def feature_l1(self, reduction="sum", average=None):
-        return self.apply_reduction(self.features.abs(), reduction=reduction, average=average)
+    def feature_l1(
+        self, reduction: Literal["sum", "mean", None] = "sum", average: Optional[bool] = None
+    ) -> torch.Tensor:
+        return self.apply_reduction(self.features.abs(), reduction=reduction, average=average)  # type: ignore[no-untyped-call,no-any-return]
 
-    def regularizer(self, reduction="sum", average=None):
-        return self.feature_l1(reduction=reduction, average=average) * self.feature_reg_weight
+    def regularizer(
+        self, reduction: Literal["sum", "mean", None] = "sum", average: Optional[bool] = None
+    ) -> torch.Tensor:
+        return self.feature_l1(reduction=reduction, average=average) * self.feature_reg_weight  # type: ignore[no-any-return]
 
-    def forward(self, x, shift=None):
+    def forward(self, x: torch.Tensor, shift: Optional[Any] = None) -> torch.Tensor:
         attention = self.attention(x)
         b, c, w, h = attention.shape
         attention = F.softmax(attention.view(b, c, -1), dim=-1).view(b, c, w, h)
-        y = torch.einsum("bnwh,bcwh->bcn", attention, x)
-        y = torch.einsum("bcn,nc->bn", y, self.features)
+        y = torch.einsum("bnwh,bcwh->bcn", attention, x)  # type: ignore[attr-defined]
+        y = torch.einsum("bcn,nc->bn", y, self.features)  # type: ignore[attr-defined]
         if self.bias is not None:
             y = y + self.bias
-        return y
+        return y  # type: ignore[no-any-return]
```

Comment:
Is this an ignore, which is a result of ignoring einsum? Then maybe it would be better to set line 89 to:
```
        y: torch.Tensor = torch.einsum("bnwh,bcwh->bcn", attention, x)  # type: ignore[attr-defined]
 ```
Then we know what the type of y should be and we don't need to ignore when returning

Reply:
Ah, yes that is better.